frmts/gtiff/tif_jxl.c

```diff
@@ -326,6 +411,31 @@ JXLPreDecode(TIFF* tif, uint16_t s)
             JxlDecoderReleaseInput(sp->decoder);
             return 0;
         }
+        if( nFirstExtraChannel < info.num_extra_channels ){
+            //first reorder the main buffer
+            int nMainChannels = bAlphaEmbedded?info.num_color_channels+1:info.num_color_channels;
+            int mainPixSize = nMainChannels*nBytesPerSample;
+            int fullPixSize = td->td_samplesperpixel*nBytesPerSample;
+            unsigned int outOff = sp->uncompressed_size - fullPixSize;
+            int inOff = main_buffer_size - mainPixSize;
+            for(;inOff>=0;inOff-=mainPixSize,outOff-=fullPixSize)
+            {
+                memcpy(sp->uncompressed_buffer+outOff,sp->uncompressed_buffer+inOff,mainPixSize);
```

Comment:
there might be some perf hits to have memcpy() in a loop with the size argument being a variable and not a constant. Given that mainPixSize will generally be in (1, 2, 3, 4, 6, 8) range   ( nbands = [1,3,4] * nBytesPerSample = [1,2] ) ), it might be worth to have  particularized for loops with mainPixSize hardcoded to those values (and the current loop as a fallback)
Same for below loop for the extra channels